contrib_bots/lib/giphy/giphy.py

```diff
@@ -0,0 +1,116 @@
+# To use this plugin, you need to set up the Giphy API key for this bot in
+# zulip_giphy_config.py
+
+from __future__ import absolute_import
+from . import zulip_giphy_config as config
+import requests
+import logging
+
+class GiphyHandler(object):
+    '''
+    This plugin posts a GIF in response to the keywords provided by the user.
+    Images are provided by Giphy, through the public API.
+    The bot looks for messages starting with "@giphy" or @mention of the bot
+    and responds with a message with the GIF based on provided keywords.
+    It also responds to private messages.
+    '''
+    def usage(self):
+        return '''
+            This plugin allows users to post GIFs provided by Giphy.
+            Users should preface keywords with "@giphy" or the Giphy-bot @mention.
+            The bot responds also to private messages.
+            '''
+
+    def get_url_gif_giphy(self, keyword, api_key):
+        # Return a URL for a Giphy GIF based on keywords given.
+        # In case of error, e.g. failure to fetch a GIF URL, it will
+        # return a number.
+        query = {'s': keyword,
+                 'api_key': api_key}
+        try:
+            data = requests.get('http://api.giphy.com/v1/gifs/translate', params=query)
+        except requests.exceptions.ConnectionError as e:  # Usually triggered by bad connection.
+            logging.warning(e)
+            return 1
+
+        if not data.status_code == 200:
+            return 1
+
+        search_status = data.json()['meta']['status']
+        if not search_status == 200:
+            return 1
+
+        try:
+            gif_url = data.json()['data']['images']['original']['url']
+        except TypeError:  # Usually triggered by no result in Giphy.
```

Comment:
Actually when it returns no result, the dictionary will read by python as `list` so `KeyError` type will not catch the `TypeError` exception here.
```
Traceback (most recent call last):
  File "run.py", line 124, in <module>
    run()
  File "run.py", line 120, in run
    quiet=options.quiet
  File "run.py", line 80, in run_message_handler_for_bot
    client.call_on_each_message(handle_message)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/zulip/__init__.py", line 486, in call_on_each_message
    self.call_on_each_event(event_callback, ['message'])
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/zulip/__init__.py", line 478, in call_on_each_event
    callback(event)
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/zulip/__init__.py", line 485, in event_callback
    callback(event['message'])
  File "run.py", line 76, in handle_message
    state_handler=state_handler
  File "/srv/zulip/contrib_bots/lib/giphy.py", line 88, in handle_message
    gif_url = self.get_url_gif_giphy(keyword, config.GIPHY_API_KEY)
  File "/srv/zulip/contrib_bots/lib/giphy.py", line 46, in get_url_gif_giphy
    gif_url = data.json()['data']['images']['original']['url']
TypeError: list indices must be integers, not str
```

```py
>>> data = requests.get('http://api.giphy.com/v1/gifs/translate?s=zulip&api_key=dc6zaTOxFJmzC')
>>> pprint.pprint(data.json())
{u'data': [],
 u'meta': {u'msg': u'OK',
           u'response_id': u'5852b48f49fcdd3785ade1ad',
           u'status': 200}}
>>> 
```